### ISSUE TITLE
Logging messages that exceed the maximum size

### ISSUE THREAD

**Issue by mcboothy** [labrixdigital/cloudwatch#3]
boto3 put_log_events has a maximum length in bytes (1,048,576 bytes)

see here 
[CloudWatchLogs.Client.put_log_events](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/logs.html#CloudWatchLogs.Client.put_log_events)

This causes an exception if an attempt to log a message that exceeds the limit.

**Comment by ernestomi** [comment]
Hi @mcboothy! 

Thanks for opening the issue. Could you elaborate on what you propose as a solution? Or how have you managed to overcome this limitation outside of this library?

Unfortunatelly, we cant change the maximum length limitation imposed by AWS. So I am trying to understand what would solve your issue. From what I can see:

- Splitting the message is not really an option since you are making an assumption for the user. If this was a good design, AWS would have implemented themselves via boto3
- Changing the error message and type could be interesting, but I would open a new issue since this would need to cover all possible boto3 errors

**Comment by computercolin** [comment]
First, this a a great log handler!

Second, I think software should have good defaults. When presented with large log entries (ex: dumping an http request) that exceeds Cloudwatch event size limits, this log handler will crash your software. Crashing is not a good default.

I have added a PR for change that adds truncation: https://github.com/labrixdigital/cloudwatch/pull/5

In some cases you might want to split the log entry into multiple parts. If someone is interested, I would suggest you extend this commit and give this handler two different modes 1) truncate 2) split log message.

**Comment by mcboothy** [comment]
Hi @ernestomi 

The problem became apparent when we tried to log stack traces of which some of them were far too large to fit into Cloudwatch in a single event. 

The main problem was that this error initially stopped the service running on an EC2 and thus the cloudwatch logs appeared to stop as the error wasn’t logged. 

Only logging onto the instance and inspecting the service log showed the true error but wasn’t easily visible at first as the errors appeared to be due to logging to cloudwatch.

The not so elegant workaround was to wrap the service in a try / catch to simply log the message.

again not an elegant solution but suited our use case ;) 

I do agree with @computercolin that a default mode to truncate with an option to split or throw the error would be a great enhancement.

Thanks

**Comment by ernestomi** [comment]
Hi @computercolin and @mcboothy 

Thanks for your responses and proposals. I think a good solution for everyone would be to add two features:

1. Introduce a parameter to control the overflow with three options: `'split'`, `'truncate'`, '`error`'. We will keep the default as `'error'` to avoid silent errors for users that already rely on this library.
2. In the case ANY uncaught error arises from AWS themselves (which is outside of our control). It will log it, and then raise the error

These changes, I hope, will address your errors whilst maintaining the expected functionality for the rest of the community.

Thanks for your contributions! I will close the issue once the fix is in place.

**Comment by computercolin** [comment]
Hi @ernestomi Thank you very much for your commitment to extend this library. I think these three modes 'split', 'truncate', and 'error' make sense. However I very much believe you should make 'truncate' the default. I understand this changes the behavior of the library, so if you need to increment the major version to denote this, you should do it.

In order to utilize the error behavior, a use would have to try, catch all calls to logging. 

```python
try:
    logging_instance.info(f"Error invoking server {http_resp_bytes}")
except:
    #  Do something?
```
 
This is cumbersome and unlikely.

The only other case where I can see a user preferring the 'error' behavior is if they would rather crash than submit a large log message to CloudWatch. However, truncation also achieves this goal. Truncation prevents submitting large, multi-megabyte messages to CloudWatch.

I really think truncating is a much safer and more useful behavior for your users. If you need to bump version numbers to communicate that behavior has changed, I would urge you to do it.


Thanks again!

**Comment by ernestomi** [comment]
@computercolin thanks for your contribution! I have used your original code and added the split feature. The new version is now released on PyPi and you appear as a contributor here on GitHub! Let me know if things dont work as expected and we can re-open

@mcboothy thanks for creating the issue in the first place and sorry it took so long to fix! Also let me know if it doesnt work as expected

**Comment by computercolin** [comment]
👏 Great. This library is wonderfully helpful and I'm glad to have error, truncate, and split options! It is a good interface. Thank you @ernestomi!
